### ISSUE TITLE
Remove orphaned states when documents are deleted

### ISSUE THREAD

**Issue by daun** [loupe-php/loupe#112]
- DELETE all states in `state_set` that are not referenced in either `terms.state` or `prefixes_terms.state`
- Came up in discussion of #110

**Comment by daun** [comment]
Depending on how the state integer is derived, could one use a recursive function to get the unused states? I've tried this for a while yesterday, assuming that somehow the used states can be tracked back from the final integer using modulo. Then I gave up 🤠

```sql
WITH RECURSIVE state_parts(original_state, remaining_state) AS (
  -- Initial state
  SELECT state, state FROM terms
  UNION ALL
  -- Keep dividing by alphabet size (4)
  SELECT original_state, remaining_state / 4
  FROM state_parts
  WHERE remaining_state > 0
)
SELECT DISTINCT original_state
FROM state_parts
  -- Check if 3 was used anywhere to derive that particular state int
WHERE remaining_state % 4 = 3
```

**Comment by Toflar** [comment]
Hmm, not sure myself anymore. So any of the terms that were unique to the document that you delete caused intermediate states which then also influence the states of other documents.
Example with index length `6` and alphabet size `4` and the `Utf8Alphabet` which is the default:
Indexing `foobar` causes 6 states:
- 3
- 16
- 68
- 275
- 1102
- 4411 (also assigned to `foobar` as the term itself

Now if I search for `foobaz` with a distance of 2, I would get the following matching states:
- 1102
- 4411

So `foobaz` does match, which is correct.
If I delete `foobar` now, the 6 states currently remain present which means that searching for `foobaz` would still return the 2 states.

However, this is not a problem because the document terms do not exist anymore and would not match. So what it will do is cause a search for term states that do not exist anymore. That's not a big deal, though. No false-positives in this case.

However, if you change the config, the entire state set must be updated.
Or in other words, I don't really know if we should even do anything here until the first person encounters an issue and we have a use case 😄

**Comment by daun** [comment]
Sounds okay if there's no false positives! The engine [already reports as "needs reindexing"](https://github.com/loupe-php/loupe/blob/8a9aae9bc46cc470483e82631024567d528664d2/src/Internal/Engine.php#L207) if the alphabet and index size change, so I'd expect users to have their own logic for re-indexing in that case.

**Comment by daun** [comment]
I see. What about truncating the `state_set` table and dumping the cache to disk when deleting all documents, along with all documents and terms? It's not perfect, but in that case we're pretty sure what to remove (everything). That doesn't solve atomic deletes, but it's a start.

**Comment by daun** [comment]
Solved via #117
